### PR TITLE
Adding tpu capabilities to train.py

### PR DESCRIPTION
**Pull request by taylanbil**
This PR will create the `tpu` branch in our forked fairseq repo.

Addressed Soumith's comments from https://github.com/pytorch-tpu/fairseq/pull/7

main file with changes is train.py. This file now has the bulk of `pytorch-tpu/examples/fairseq_train_tpu.py` which is going away once this pr is checked in.

Other files that are changed contain changes from pytorch-tpu/examples/fairseq_train_tpu.py as well.

### CODE REVIEW

**Comment by taylanbil** [comment]
I tested this by 
* training+validating 2 epochs on the validation set.
* Saving checkpoint.
* Loading checkpoint and doing 1 more epoch
* Saving that checkpoint.

File: train.py [hunk 1]
```diff
@@ -264,42 +277,317 @@ def distributed_main(i, args, start_rank=0):
...
+    parser.add_argument('--metrics_debug', action='store_true')
+    parser.add_argument('--use_gpu', action='store_true')
+    args = options.parse_args_and_arch(parser)
 
-    if args.distributed_init_method is not None:
-        # distributed training
-        if torch.cuda.device_count() > 1 and not args.distributed_no_spawn:
-            start_rank = args.distributed_rank
-            args.distributed_rank = None  # assign automatically
+    if args.use_gpu:
```
Comment (dlibenzi) [inline train.py]: This function is pretty big and hard to follow.
I'd honestly have something like:
```Python
def cli_main():
  # Common init code
  ...
  if gpu:
    main_gpu()
  else:
    main_tpu()
```

**Review by dlibenzi — CHANGES_REQUESTED** [review]

File: fairseq/data/data_utils.py [hunk 2]
```diff
@@ -24,9 +24,25 @@ def infer_language_pair(path):
...
+    if input_shapes is None:
+        return max(v.size(0) for v in values)
+    for batch_size, padlen in input_shapes:
+        if len(values) == batch_size:
+            return padlen
+    else:
+        raise IndexError(
+            'Encountered values with invalid length {}, input shapes were {}'
+            .format(len(values), input_shapes)
+        )
```
Comment (zcain117) [inline fairseq/data/data_utils.py:37]: This indenting is weird to me: why is the `for` loop unindented but in between the `if` and `else`? I'd probably put the `for` loop inside the `else` block unless there's some style point here I'm missing

File: train.py [hunk 3]
```diff
@@ -302,5 +521,83 @@ def cli_main():
...
+    if nonnull_gpu_input_shape_args:
+      errmsg = (
+          'On TPUs, please control input shapes '
+          'using `--input_shapes`. Any non-null arg in {} will trigger'
+          ' this error.'
+      ).format(gpu_input_shape_args)
+      raise RuntimeError(errmsg)
+
+    args.input_shapes = parse_input_shapes(args.input_shapes)
+    # XXX (taylanbil): do we ever have more than 2 dimensions in fairseq?
```
Comment (zcain117) [inline train.py:580]: Just confirming this is still an open question

File: fairseq/data/data_utils.py (hunk 2 above)
Comment (taylanbil) [reply fairseq/data/data_utils.py:37]: for/else is a python thing. else is executed if you don't early eexit out of the for loop. here, len(values) _must_ be inside input_shapes and we must early return, otherwise we raise error. I can get rid of the else, it would have the same effect, wdyt?

File: train.py (hunk 3 above)
Comment (taylanbil) [reply train.py:580]: it is.

File: fairseq/data/data_utils.py [hunk 4]
```diff
@@ -187,8 +203,24 @@ def is_batch_full(num_tokens):
 
         batch.append(idx)
 
-    # if len(batch) > 0:
-    #     yield batch
+    if len(batch) > 0:
+        yield batch
+
+
+def batch_by_size_tpu(
```
Comment (dlibenzi) [inline fairseq/data/data_utils.py:210]: Make sure we follow the style of native code.
Not saying this isn't ... just be sure we are not doing "our own" style.

**Review by dlibenzi — APPROVED** [review]

File: fairseq/data/data_utils.py (hunk 4 above)
Comment (taylanbil) [reply fairseq/data/data_utils.py:210]: yep, I'm familiar w/ FB formatting and I manually checked, it would be better if we had a formatter. I asked around but nobody gave me anything.

File: fairseq/data/data_utils.py (hunk 2 above)
Comment (zcain117) [reply fairseq/data/data_utils.py:37]: cool, TIL. Looks good as-is in that case

**Review by zcain117 — APPROVED** [review]
